patch/llpcPatchCheckShaderCache.cpp

```diff
@@ -0,0 +1,161 @@
+} // anonymous
+
+// =====================================================================================================================
+PatchCheckShaderCache::PatchCheckShaderCache()
+    :
+    Patch(ID)
+{
+    initializePipelineShadersPass(*PassRegistry::getPassRegistry());
+    initializePatchCheckShaderCachePass(*PassRegistry::getPassRegistry());
+}
+
+// =====================================================================================================================
+// Executes this LLVM patching pass on the specified LLVM module.
+bool PatchCheckShaderCache::runOnModule(
+    Module& module)  // [in,out] LLVM module to be run on
+{
+    LLVM_DEBUG(dbgs() << "Run the pass Patch-Check-Shader-Cache\n");
+
+    if (m_callbackFunc == nullptr)
+    {
+        // No shader cache in use.
+        return false;
+    }
+
+    Patch::Init(&module);
+
+    MD5::MD5Result inOutUsageHashes[ShaderStageGfxCount] = {};
+    ArrayRef<uint8_t> inOutUsageHashValues[ShaderStageGfxCount] = {};
+    auto stageMask = m_pContext->GetShaderStageMask();
+
+    // Build input/output layout hash per shader stage
+    for (auto stage = ShaderStageVertex; stage < ShaderStageGfxCount; stage = static_cast<ShaderStage>(stage + 1))
+    {
+        if ((stageMask & ShaderStageToMask(stage)) == 0)
+        {
+            continue;
+        }
+
+        auto pResUsage = m_pContext->GetShaderResourceUsage(stage);
+        MD5 hasher;
+
+        // Update input/output usage
+        UpdateHashForMap(pResUsage->inOutUsage.inputLocMap, &hasher);
+        UpdateHashForMap(pResUsage->inOutUsage.outputLocMap, &hasher);
+        UpdateHashForMap(pResUsage->inOutUsage.perPatchInputLocMap, &hasher);
+        UpdateHashForMap(pResUsage->inOutUsage.perPatchOutputLocMap, &hasher);
+        UpdateHashForMap(pResUsage->inOutUsage.builtInInputLocMap, &hasher);
+        UpdateHashForMap(pResUsage->inOutUsage.builtInOutputLocMap, &hasher);
+        UpdateHashForMap(pResUsage->inOutUsage.perPatchBuiltInInputLocMap, &hasher);
+        UpdateHashForMap(pResUsage->inOutUsage.perPatchBuiltInOutputLocMap, &hasher);
+
+        if (stage == ShaderStageGeometry)
+        {
+            // NOTE: For geometry shader, copy shader will use this special map info (from built-in outputs to
+            // locations of generic outputs). We have to add it to shader hash calculation.
+            UpdateHashForMap(pResUsage->inOutUsage.gs.builtInOutLocs, &hasher);
+        }
+
+        // Store the result of the hash for this shader stage.
+        hasher.final(inOutUsageHashes[stage]);
+        inOutUsageHashValues[stage] = inOutUsageHashes[stage].Bytes;
+    }
+
+    // Ask callback function if it wants to remove any shader stages.
+    uint32_t modifiedStageMask = m_callbackFunc(&module, stageMask, inOutUsageHashValues);
+    if (modifiedStageMask == stageMask)
+    {
+        return false;
+    }
+
+    // "Remove" a shader stage by making its entry-point function internal, so it gets removed later.
```

Comment:
have you compare the compile performance with original code?

Reply:
On the first version of this change, I did not time the compile performance, but I did check that it got the same cache hits compiling all the pipelines from a representative game in the same amdllpc with the option to enable the cache. I have not done that check on this latest version yet, but I will do.

Reply:
I have now confirmed that it gets exactly the same cache hits and has the same compile performance on all graphics pipelines from a representative game.